### ISSUE TITLE
Idea: Simple and good Integration with ASP.NET Core

### ISSUE THREAD

**Issue by altmann** [altmann/FluentResults#149]
If you are interested like the issue :)

**Comment by eltoncezar** [comment]
I've created something very simple to deal with Results and HTTP Status Codes, please take a look:
https://github.com/eltoncezar/FluentResults.StatusCodes

**Comment by altmann** [comment]
Cool. I was not aware that something similar already exists. I had a look at it - I think that the new aspnetcore integration package will be similar to your code but in total very different. I spiked already a little bit. :)

**Comment by altmann** [comment]
I pushed currently my first version of the aspnetcore integation. https://github.com/altmann/FluentResults/tree/features/AspNetCoreIntegration/src/FluentResults.Samples.WebHost/Controllers

Public APIs
- ToActionResult(...) to transform a Result / Result<T> to a HttpResponse [example](https://github.com/altmann/FluentResults/blob/features/AspNetCoreIntegration/src/FluentResults.Samples.WebHost/Controllers/WeatherForecastWithControllerSpecificProfileController.cs)
- Optional custom transformation behaviour per Controller 
  - Your custom domain/technical errors can be translated to different http responses/http statuscodes
  - [example](https://github.com/altmann/FluentResults/blob/features/AspNetCoreIntegration/src/FluentResults.Samples.WebHost/Controllers/CustomAspNetCoreResultEndpointProfile.cs)
- Possible to overwrite the default transformation behavior [sample](https://github.com/altmann/FluentResults/blob/features/AspNetCoreIntegration/src/FluentResults.Samples.WebHost/Program.cs)

Feedback welcome!

**Comment by awaters-meltin** [comment]
I recently went down the rabbit hole with this and come up with something somewhat similar to @eltoncezar w.r.t. custom error types (but errors just implementing an `IStatusCodeError`). But rather than using a `Validate` method or `.ToActionResult()` extension method I'm using a `ResultFilter` (I very quickly got tired of calling `Validate` and/or `.ToActionResult()` everywhere).

This might be a bit too much for the library's purposes, but I wanted all the `Result` info (stack traces etc.) in the Asp .NET standard `ProblemDetails` response (it's an internal project so better for logging).

Here's a code dump for ya lol - cleaned it up a bit, but still using various internal extension methods, but the general gist should be understandable:
```C#
builder.Services.AddControllers(options => { options.Filters.Add<ResultTypeFilter>(); })

// ...

internal class ResultTypeFilter : IAsyncResultFilter
{
    public async Task OnResultExecutionAsync(ResultExecutingContext context, ResultExecutionDelegate next)
    {
        if (context.Result is ObjectResult { Value: IResultBase result })
            context.Result = ToActionResult(result);

        await next();
    }

    private static ActionResult ToActionResult(IResultBase result)
    {
        if (result.IsSuccess)
            return new OkObjectResult((result as IResult<object>)?.Value);

        IEnumerable<string> ExceptionToStringList(Exception e)
            => new[] { $"{e.GetType().Name}: {e.Message}" }.Concat((e.StackTrace ?? "").Split(Environment.NewLine));

        var awsException = result.GetException<AmazonServiceException>();
        var statusCode = result.Errors.OfType<IStatusCodeError>().FirstOrDefault()?.StatusCode ??
                         (int?) awsException?.StatusCode;

        var e = awsException ?? result.GetException();
        var problemDetails = new ProblemDetails
        {
            Title = e?.Message ?? result.Errors.FirstOrDefault()?.Message ?? "Unknown Error",
            Status = statusCode ?? (e == null ? 400 : 500),
            Detail = result.Reasons
                           .Select(reason => reason.Message)
                           .JoinToString(sep: ", ", prefix: "[", suffix: "]", affixEmpty: false, affixSingle: false),
            Extensions =
            {
                ["errors"] =
                    result.Errors
                          .OfType<Error>() // using "Error" type for automatic JSON conversion
                          .Select(
                              err =>
                              {
                                  if (err is IExceptionalError exErr)
                                      err.Metadata["_stackTrace"] = ExceptionToStringList(exErr.Exception);

                                  return err;
                              }
                          )
            }
        };

        return new ObjectResult(problemDetails) { StatusCode = problemDetails.Status };
    }
}
```

Then my controller endpoints are:
```C#
    [HttpPost, RemoteStorageDevRoute("credentials")]
    public async Task<Result<CredentialsRecord>> FetchCredentials([FromBody] StoreCredentials credentials)
        => await _store.FetchCredentials(credentials.Key);
```

The one downside to this was the Swagger generator needed to be tweaked to handle the `Result<T>` types a bit better (using a `IOperationFilter` on SwaggerGen), but other than that it's been working quite well.

**Comment by altmann** [comment]
@awaters-meltin Thank you for your comment. I also played around a bit with an aspnetcore filter but then I deleted the code. 

This biggest disadvantage of filters is the intransparency - the controller api is lying - in background the return value (=result object) of the controller is translated into something else. You write it already, swagger need because of that some custom configuration to work correctly. The benefit is that you don't have to write ToActionResult() again and again. 

Do other people have some opinions about this topic?

**Comment by altmann** [comment]
Is the usage of the [ProblemDetails](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.problemdetails?view=aspnetcore-7.0) class a MUST-HAVE in the default logic of FluentResults.Extensions.AspNetCore? For sure the default logic can be overwritten.

**Comment by awaters-meltin** [comment]
If they're trying to follow the official RFCs etc. (https://www.rfc-editor.org/rfc/rfc7807) then they'd want the `ProblemDetails`, but some companies love to roll their own error format (of course lol).

So I don't know about a "must have", but I don't think there's a reason not to. 🤷

**Comment by altmann** [comment]
Haha, in my current project in the job we also have a custom error model. ;)

I will have a look at the ProblemDetails class in the next days and I hope I can release a first preview package of the aspnetcore integration until the end of october. Thanks for your comment.

**Comment by altmann** [comment]
> This will be an excellent addition.
> 
> It would be good if `ToActionResult` on `Result<T>` returned `ActionResult<T>` rather than `ActionResult`.

I tried it but then it is not possible to change the resulting dto in the background by using the extension points.

**Comment by altmann** [comment]
Thanks for all the feedback. 

Because I'am out of time I released today the [v0.1.0](https://www.nuget.org/packages/FluentResults.Extensions.AspNetCore/0.1.0). Its on nuget, so you should be able to use it. If you use it read the [documentation](https://github.com/altmann/FluentResults/wiki/Returning-Result-Objects-from-ASP.NET-Core-Controller) 

If you have feedback to the functionality/api or the documentation please let me know.

**Comment by eltoncezar** [comment]
@altmann would you give us permission to edit wiki pages?

**Comment by altmann** [comment]
I don't know how to so it. Maybe you have to clone the wiki repo and make a pr https://github.com/altmann/FluentResults.wiki.git

**Comment by saborrie** [comment]
Does this also cover minimal APIs too? Microsoft does not use ActionResults for those, instead you are supposed to return a `Microsoft.AspNetCore.Http.IResult`, usually from the factory `Microsoft.AspNetCore.Http.Results`, e.g. 

```csharp
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

app.MapGet("/", () => Results.BadRequest("Some error message."));

app.Run();
```

**Comment by billhong-just** [comment]
> Does this also cover minimal APIs too? Microsoft does not use ActionResults for those, instead you are supposed to return a `Microsoft.AspNetCore.Http.IResult`, usually from the factory `Microsoft.AspNetCore.Http.Results`, e.g.
> 
> ```cs
> var builder = WebApplication.CreateBuilder(args);
> var app = builder.Build();
> 
> app.MapGet("/", () => Results.BadRequest("Some error message."));
> 
> app.Run();
> ```

@altmann Do you have any thoughts on this?

**Comment by angusbreno** [comment]
Hey guys take a look:

https://github.com/ElysiumLabs/FluentProblemDetails/

**Comment by PeterKnealeCMRI** [comment]
I dont believe `FluentResults.Extensions.AspNetCore` v0.2.0 works with minimal apis. 
They return `IResult` from `Microsoft.AspNetCore.Http` as opposed to `ActionResult` in `Microsoft.AspNetCore.Mvc` as mentioned above.

Just as an experiment I returned the ActionResult from a minimal api using `.ToActionResult()` populated the body with a `Microsoft.AspNetCore.Mvc.BadRequestObjectResult` that serialised and was stamped with a HTTP 200 response.

very confusing initally :)

```cs
var result = await mediator.Send(command);
return result.ToActionResult();
```

Response body: 
```json
{
"value":[{"message":"The term ABCa0b47c93-600e-417a-82ab-83ec245eba5e already exists."}],
"formatters":[],
"contentTypes":[],
"declaredType":null,
"statusCode":400
}
```

StatusCode: `OK`

**Comment by angusbreno** [comment]
@PeterKnealeCMRI 

Thanks for reporting this and for the example, it really helped to clarify the scenario.

The behavior you observed is expected when using Minimal APIs and returning an ActionResult.

In Minimal APIs, the runtime expects handlers to return Microsoft.AspNetCore.Http.IResult.
If an ActionResult (from Microsoft.AspNetCore.Mvc) is returned instead, it is treated as a regular object and serialized, which results in a 200 OK response and a JSON payload containing the internal properties of BadRequestObjectResult.

This package supports Minimal APIs through an endpoint filter, not by returning ActionResult directly.

The correct usage with Minimal APIs is to attach the provided endpoint filter to the route (or route group), and return the FluentResults.Result directly from the handler.

Example:

app.MapPost("/terms", async (IMediator mediator, CreateTermCommand command) =>
{
    return await mediator.Send(command);
})
.AddEndpointFilter<FluentResults.Extensions.AspNetCore.Http.ResultEndpointFilter>();

With this setup:
	•	successful results are mapped to 200 OK automatically
	•	failed results are mapped to a proper ProblemDetails response with the correct HTTP status code

In other words, ToActionResult() is intended for MVC controllers, while Minimal APIs should use the provided ResultEndpointFilter.

I will also review the package in the next days and apply a few mandatory improvements to make this integration easier and more explicit (for example, a simpler registration extension and clearer documentation for Minimal APIs).
